src/trace.rs

```diff
@@ -71,120 +91,150 @@ pub fn init(color: bool, json: bool, levels: &str) {
         let rate_limited = RateLimitedLayer::new(formatter);
         let subscriber = subscriber.with(rate_limited.with_filter(fmt_filter));
 
-        let subscriber = BroadcastSubscriber { subscriber };
         let _ = subscriber.try_init();
     }
 }
 
 #[cfg(test)]
-pub fn reset_early_buffer() {
-    *early_buffer() = Some(Vec::new());
+pub fn reset_early_buffer() -> Option<Vec<LogEvent>> {
+    get_early_buffer().replace(Vec::new())
 }
 
-fn early_buffer() -> MutexGuard<'static, Option<Vec<LogEvent>>> {
+fn get_early_buffer() -> MutexGuard<'static, Option<Vec<LogEvent>>> {
     BUFFER
         .get()
         .expect("Internal logs buffer not initialized")
         .lock()
         .expect("Couldn't acquire lock on internal logs buffer")
 }
 
-pub fn stop_buffering() {
-    *early_buffer() = None;
+fn try_buffer_event(event: LogEvent) -> Option<LogEvent> {
+    if SHOULD_BUFFER.load(Ordering::Acquire) {
+        if let Some(buffer) = get_early_buffer().as_mut() {
+            buffer.push(event);
+            return None;
+        }
+    }
+
+    Some(event)
 }
 
-/// Gets the current [`Span`].
-pub fn current_span() -> Span {
-    Span::current()
+fn consume_early_buffer() -> Vec<LogEvent> {
+    get_early_buffer()
+        .take()
+        .expect("early buffer was already consumed")
 }
 
-pub struct TraceSubscription {
-    pub buffer: Vec<LogEvent>,
-    pub receiver: Receiver<LogEvent>,
+/// Creates a trace sender for sending internal log events.
+fn get_trace_sender() -> &'static broadcast::Sender<LogEvent> {
+    SENDER.get_or_init(|| broadcast::channel(99).0)
 }
 
-pub fn subscribe() -> TraceSubscription {
-    let buffer = match early_buffer().as_mut() {
-        Some(buffer) => buffer.drain(..).collect(),
-        None => Vec::new(),
-    };
-    let receiver = SENDER.get_or_init(|| broadcast::channel(99).0).subscribe();
-    TraceSubscription { buffer, receiver }
+/// Creates a trace receiver that receives internal log events.
+fn get_trace_receiver() -> broadcast::Receiver<LogEvent> {
+    get_trace_sender().subscribe()
 }
 
-struct BroadcastSubscriber<S> {
-    subscriber: S,
+fn get_trace_subscriber_list() -> MutexGuard<'static, Option<Vec<oneshot::Sender<Vec<LogEvent>>>>> {
+    SUBSCRIBERS
+        .get_or_init(|| Mutex::new(Some(Vec::new())))
+        .lock()
+        .expect("poisoned locks are dumb")
```

Comment:
Checks out.

Reply:
I was surprised we didn't already pull in `parking_lot` as a dependency, which doesn't have poisonable locks... so this was mostly me just getting the boilerplate out of the way, but I would probably switch to `parking_lot` if this PR looked good overall.